### PR TITLE
Add prefix support for environment variables

### LINKED ISSUE #3450: Add prefix support for environment variables
Currently environment variables get bound to `@Configuration` properties directly (for example `SERVER_PORT` to `Server.port`). This can cause problems when trying to set `ignoreInvalidFields` or if you want to run two different boot apps as the same user.

@gregturn suggested that we consider supporting a prefix when binding. For example:

```
springApplication.setEnvironmentPrefix("MYAPP_");
...
```

Which would then bind `MYAPP_SERVER_PORT` rather than `SERVER_PORT`.

### PR DESCRIPTION
**Pull request by lucassaldanha**
Creates an environmentPrefix property that is used to filter the
SystemEnvironmentPropertySource's map of environment
variables and create a new one only with the prefixed properties.
This new map is used on a new SystemEnvironmentPropertySource
that replaces the former one.

Fixes #3450

### CODE REVIEW

**Comment by lucassaldanha** [comment]
@philwebb I didn't understand why the build failed. Looks like some error in Travis CI worker... Do you know what happened?

**Comment by snicoll** [comment]
I don't think that's what we had in mind. We should change the existing `PropertySource` as anybody may want to look it up manually (either via the `Environment` or `@Value`). 

We thought of applying that constraint for **binding** so that not the full OS environment is available as it may actually break it (see the discussion on the original issue).

**Comment by lucassaldanha** [comment]
@snicoll Ok! I'll adjust the code to match the behavior you described.

File: spring-boot/src/main/java/org/springframework/boot/SpringApplication.java [hunk 1]
```diff
@@ -843,6 +845,17 @@ public void setEnvironment(ConfigurableEnvironment environment) {
 	}
 
 	/**
+	 +	 * Sets the environment variables prefix that will be used to filter those variables
+	 +	 * for this application.
+	 +	 * @param environmentPrefix the prefix
+	 +	 */
+	public void setEnvironmentPrefix(String environmentPrefix) {
```
Comment (lucassaldanha) [inline spring-boot/src/main/java/org/springframework/boot/SpringApplication.java]: @snicoll I'm not sure if this is the best way to configure the environment prefix. I'd like some opinions on that.

**Comment by lucassaldanha** [comment]
@snicoll I believe that now it does what is is expected to do (considering the prefix only on binding).

I made a line note about the way that the environment prefix is being configured. Because I'm not sure if it's a good idea using the system properties to do that. Need some opinions.

**Comment by snicoll** [comment]
@lucassaldanha this PR was automatically closed because you made it from master and your changes went away when you push force on your fork. If you still want to contribute this change, please create a PR from a branch and not master. Thanks.

**Pull request closed without merging**